diff --git a/main.c b/main.c
index 8edb2e1..35cdb3c 100644
--- a/main.c
+++ b/main.c
@@ -492,8 +492,10 @@ int main
     }
   }
 
+#if 0
   if (getuid() && !client_only)
     LOG_FATAL("Not superuser");
+#endif
 
   /* Turn into a daemon */
   if (!nofork) {
@@ -530,8 +532,10 @@ int main
     user = CNF_GetUser();
 
   pw = getpwnam(user);
+#if 0
   if (!pw)
     LOG_FATAL("Could not get user/group ID of %s", user);
+#endif
 
   /* Create directories for sockets, log files, and dump files */
   CNF_CreateDirs(pw->pw_uid, pw->pw_gid);
@@ -566,9 +570,11 @@ int main
     SYS_LockMemory();
   }
 
+#if 0




   /* Drop root privileges if the specified user has a non-zero UID */
   if (!geteuid() && (pw->pw_uid || pw->pw_gid))
     SYS_DropRoot(pw->pw_uid, pw->pw_gid);
+#endif
 
   REF_Initialise();
   SST_Initialise();
